### ISSUE TITLE
PCA on gpu hangs

### ISSUE THREAD

**Issue by fcharras** [intel/scikit-learn-intelex#1534]
**Describe the bug**

I'm setting up a [benchmark project](https://github.com/soda-inria/sklearn-engine-benchmarks) to compare speed of similar algorithms with different backends (results are browsable in [this spreadsheet](https://docs.google.com/spreadsheets/d/1te_3jY6vI4wo3-V7xbmWQai5Mdh5plWdLst2ox0wuLM)), and I want to include `scikit-learn-intelex` performance whenever it's possible. Now working on a [PCA benchmark](https://github.com/soda-inria/sklearn-engine-benchmarks/pull/2), for scikit-learn-intelex PCA works fine on CPU but on GPU it just seems to hang. 

**To Reproduce**

```python
import numpy as np
import dpctl
import dpctl.tensor as dpt
from sklearnex.decomposition import PCA

rng = np.random.default_rng(123)
X = rng.random((2_000_000,100), dtype=np.float32)

device=dpctl.SyclDevice("gpu:level_zero")
X = dpt.asarray(X, copy=True, device=device)

pca = PCA(n_components=10, svd_solver="full")
# Also hangs
# pca = PCA(n_components=10)
pca.fit(X)
```

**Expected behavior**
Completes in a reasonable amount of time (seconds at most)

**Output/Screenshots**
No output, it just hangs.

**Environment:**

`scikit-learn-intelex==2023.2.1` installed with pip, in a conda environment, on intel beta dev cloud, running on max series gpu.

**Comment by samir-nasibli** [comment]
@fcharras thank you for the report!
It seems we have an issue with dpctl support for PCA.fit. Fix will be pathced soon.
As an option you can try to use `config_context` for the GPU offload:
``` python
rng = np.random.default_rng(123)
X = rng.random((2_000_000,100), dtype=np.float32)

pca = PCA(n_components=10, svd_solver="full")
with config_context(target_offload="gpu:level_zero"):
    pca.fit(X)
 ```

**Comment by fcharras** [comment]
Thanks, will try the config context + numpy input.

Are there other estimators having the same issue ? I've been trying the gpu backend this way for other estimators (like k-means).

**Comment by samir-nasibli** [comment]
@fcharras for benchmarks I recommend to use config context + numpy.
API that supports DPNP/DPCTL just do conversions from dpctl tensor to numpy for input arrays and numpy to dpctl for outputs (for all estimators).
We are working on full integration of dpnp/dpctl without intermediate copies.

**Comment by fcharras** [comment]
I thought using on-device data with dpctl would be saving a copy. Thanks for the information.

**Comment by fcharras** [comment]
Regarding CPU implementations, what happens in those differences cases regarding CPU:
- using `target_offload="opencl:cpu"` (or passing ``target_offload="auto"` when only `opencl:cpu` is available)
- not using config context at all
- while passing as input numpy array or usm_ndarray on cpu

? does the cpu implementation uses the opencl runtime or is it different ?

**Comment by samir-nasibli** [comment]
@fcharras I recommend not to use config_context at all for CPU

**Comment by samir-nasibli** [comment]
@fcharras the original issue is fixed and fix will be able on 2024.0.1 release.
